Add MKE and CVE-2024-21626: Understanding the Potential Impact and Mitigation (#4)
Mirantis/security: advisories/0011.md

#### CVE Identifier
CVE-2024-21626
#### CVSSv3.1
8.6 (High) CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H
#### CWEs
CWE-403, CWE-668

#### Affected Product
* MKE deployment requires installation of MCR.  Older versions of MCR (<= 23.0.9) contain the vulnerability in the underlying `runc` (<= 1.1.11-rc1) component. Please see the [MCR advisory](https://github.com/Mirantis/security/blob/main/advisories/0010.md) for details.  MKE deployments with MCR <= 23.0.9 require an upgrade to MCR 23.0.9-1 or later to patch `runc` to 1.1.11-rc1.m1.

* No MKE Image Update Needed
  
   While CVE-2024-21626 might appear in vulnerability scans of your MKE images due to the dependent library, an update to MKE images is not currently required.  We understand this discrepancy can be confusing. To address this and enhance transparency, we are actively working to remove unnecessary references to the vulnerability from future MKE image scans. This update is planned for upcoming MKE releases.

#### How to check if MKE is vulnerable
1. [Download and config the client bundle](https://docs.mirantis.com/mke/3.7/ops/access-cluster/download-configure-client-bundle.html)
2. Check the `runc` version by
   ```

>    docker version --format "{{range .Server.Components}}{{if eq .Name \"runc\"}}{{.Version}}{{end}}{{end}}"

Reviewer: ```suggestion
   docker version --format '{{range .Server.Components}}{{if eq .Name "runc"}}{{.Version}}{{end}}{{end}}'
```

>    docker version --format "{{range .Server.Components}}{{if eq .Name \"runc\"}}{{.Version}}{{end}}{{end}}"

Reviewer: Wouldn't that only check the runc version installed on one of the manager nodes? Or does MKE do some magic in the Engine API proxy? How would a user check the rest of the nodes in their cluster?

Author: `docker node ls --format {{.EngineVersion}}` to check engine version.  There is no viable way other than `ssh` into nodes.

### Mirantis Kubernetes Engine (MKE) and CVE-2024-21626: Understanding the Impact and Mitigation
#### Release Date
2024-02-08
#### Overview

> This advisory addresses the impact of CVE-2024-21626, a critical vulnerability identified within runc, and its implications for the Mirantis Kubernetes Engine (MKE). Given that MKE relies heavily on the Mirantis Container Runtime (MCR), this vulnerability directly impacts its operational integrity. We aim to clarify the situation, explain the mitigation steps, and provide actionable guidance for MKE users.

Reviewer: ```suggestion
This advisory addresses the impact of CVE-2024-21626, a vulnerability identified within runc, and its implications for the Mirantis Kubernetes Engine (MKE). Given that MKE relies heavily on the Mirantis Container Runtime (MCR), this vulnerability directly impacts its operational integrity. We aim to clarify the situation, explain the mitigation steps, and provide actionable guidance for MKE users.
```
It's not a critical vulnerability; see line 11 of this very document.

### Vulnerability Information
#### CVE Identifier
CVE-2024-21626
#### CVSSv3.1
8.6 (High) CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H
#### CWEs
CWE-403, CWE-668

#### Affected Product

> * MKE deployment requires installation of MCR.  Older versions of MCR (<= 23.0.9) contain a vulnerability in the underlying runc component. Please see [the MCR advisory](https://github.com/Mirantis/security/blob/main/advisories/0010.md) for details. Any deployment of MKE with MCR version <= 23.0.9 must upgrade to MCR 23.0.9-1 or later to mitigate this risk.

Reviewer: ```suggestion
* MKE deployment requires installation of MCR.  Older versions of the MCR packages contain a vulnerable version of the runc component. Please see [the MCR advisory](https://github.com/Mirantis/security/blob/main/advisories/0010.md) for details.
```

The vulnerable versions and patching steps have some nuance that is neither correctly nor adequately captured here.

Author: While you could technically install individual packages like Docker-EE, Docker-EE-CLI, and containerd, Mirantis does not officially support such a configuration. Customers and Mirantis typically and should always treat MCR as a single, integrated product that includes all these components pre-configured and tested for compatibility. Using MCR ensures a supported and well-maintained environment, avoiding potential issues arising from manual package installation and management.

* No MKE Image Update Needed for CVE-2024-21626
  

>    While CVE-2024-21626 might appear in vulnerability scans of your MKE images due to a dependent library, an update to MKE images is not currently required. MKE's architecture mitigates the vulnerability, and there's no known exploit path within MKE. We understand this discrepancy can be confusing. To address this and enhance transparency, we're actively working to remove unnecessary references to the vulnerability from future MKE image scans. This update is planned for upcoming MKE releases.

Reviewer: > MKE's architecture mitigates the vulnerability, and there's no known exploit path within MKE.

That statement could be easily misinterpreted as "MKE protects you from the runc vulnerability." MKE _components_ are not vulnerable to this particular exploit, but MKE _does not protect users from harm_ if they ran a malicious container image with a vulnerable runc installed.

Author: Let me rewrite this paragraph

> #### How to check if MKE is vulnerable

Reviewer: But you just said MKE was not vulnerable...

Author: MKE as a product, is vulnerable due to it's dependency to MCR.  On the other hand, MKE images are not vulnerable and thus an update is not necessary.

Reviewer: ```suggestion
#### How to check if your MKE cluster is vulnerable
```

1. [Download and config the client bundle](https://docs.mirantis.com/mke/3.7/ops/access-cluster/download-configure-client-bundle.html)
2. Check the MCR version by
   ```
   docker version --format "{{range .Server.Components}}{{if eq .Name \"Engine\"}}{{.Version}}{{end}}{{end}}"
   ```

> 3. The MKE cluster is vunerable if and only if the version <= 23.0.9.

Reviewer 2: ```suggestion
3. The MKE cluster is vulnerable if and only if the version <= 23.0.9.
```

> 3. The MKE cluster is vunerable if and only if the version <= 23.0.9.

Reviewer: Check the runc version, not Engine.

Author: While I understand the root cause lies with runc, MKE customers interact directly with MCR, not runc. Therefore, focusing on MCR version checks seems more relevant.

Reviewer: Users will remain vulnerable if they upgrade their docker-ee package without upgrading their containerd.io package, as is possible (at least) on RPM distros. The _only_ way to tell for sure is to check the version of runc, **which docker version already conveniently outputs**

Author: I may add a check to `runc` and perhaps an after-upgrade check as well.  However please check https://github.com/Mirantis/security/pull/4/files#r1480921670 for my understanding of MCR as a product.

Reviewer: We want users to always treat MCR as a single integrated product but we distribute it as four separate DEB/RPM packages. There are all sorts of ways that users may get their deployments into an unsupported state, intentionally or by accident. Users do weird things, make mistakes, or they might even do everything right but fall victim to a software bug during upgrade. Users are going to use these instructions to to verify that they have applied the update correctly. Instructions which incorrectly tell users they are not vulnerable if they applied the update incorrectly are worse than useless, _they put our customers at risk_.

> #### Safely upgrading MCR to version (23.0.9-1) in a running MKE cluster
> This guide outlines the steps for upgrading MCR to version (23.0.9-1) while an MKE cluster is operational. This version of MCR has a fix for the aforementioned CVE. The process is similar to upgrading to any supported MCR version within a running MKE environment.
> 
> ##### Upgrading with Launchpad:
> 
> For a streamlined upgrade experience, follow the official instruction provided by Mirantis: [Upgrade components with Launchpad - Mirantis Kubernetes Engine](https://docs.mirantis.com/mke/3.7/launchpad/lp-upgrades.html)
> 
> ##### Upgrading manually:
> 
> If you prefer a more granular approach, refer to the Mirantis documentation for specific package installation and upgrade instructions on your chosen Linux distribution: [Install MCR on Linux distros - Mirantis Container Runtime](https://docs.mirantis.com/mcr/23.0/install/mcr-linux.html)
> 
> ##### Important notes:
> 
> * Carefully review the documentation before proceeding, as specific steps may vary depending on your cluster configuration and operating system.
> * Thoroughly test your applications after the upgrade to ensure compatibility with the new MCR version.
> * Consider backing up your cluster configuration before making any significant changes.

Reviewer: ```suggestion
```

Leave the instructions to the Engine team, please.
